elements/castbar.lua

```diff
@@ -83,443 +84,292 @@ A default texture will be applied to the StatusBar and Texture widgets if they d
+	if(element.Text) then element.Text:SetText(name) end
 	if(element.Time) then element.Time:SetText() end
 
-	local shield = element.Shield
-	if(shield and notInterruptible) then
-		shield:Show()
-	elseif(shield) then
-		shield:Hide()
-	end
-
-	local sf = element.SafeZone
-	if(sf) then
-		sf:ClearAllPoints()
-		sf:SetPoint(element:GetReverseFill() and 'LEFT' or 'RIGHT')
-		sf:SetPoint('TOP')
-		sf:SetPoint('BOTTOM')
-		updateSafeZone(element)
-	end
-
-	--[[ Callback: Castbar:PostCastStart(unit, name)
-	Called after the element has been updated upon a spell cast start.
+	local safeZone = element.SafeZone
+	if(safeZone) then
+		local isHoriz = element:GetOrientation() == 'HORIZONTAL'
 
-	* self - the Castbar widget
-	* unit - unit for which the update has been triggered (string)
-	* name - name of the spell being cast (string)
-	--]]
-	if(element.PostCastStart) then
-		element:PostCastStart(unit, name)
-	end
-	element:Show()
-end
+		safeZone:ClearAllPoints()
+		safeZone:SetPoint(isHoriz and 'TOP' or 'LEFT')
+		safeZone:SetPoint(isHoriz and 'BOTTOM' or 'RIGHT')
 
-local function UNIT_SPELLCAST_FAILED(self, event, unit, castID)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+		if(element.casting) then
+			safeZone:SetPoint(element:GetReverseFill() and (isHoriz and 'LEFT' or 'BOTTOM') or (isHoriz and 'RIGHT' or 'TOP'))
+		else
+			safeZone:SetPoint(element:GetReverseFill() and (isHoriz and 'RIGHT' or 'TOP') or (isHoriz and 'LEFT' or 'BOTTOM'))
+		end
 
-	local element = self.Castbar
-	if(element.castID ~= castID) then
-		return
-	end
+		local ratio = (select(4, GetNetStats()) / 1000) / element.max
+		if(ratio > 1) then
+			ratio = 1
+		end
 
-	local text = element.Text
-	if(text) then
-		text:SetText(FAILED)
+		safeZone[isHoriz and 'SetWidth' or 'SetHeight'](safeZone, element[isHoriz and 'GetWidth' or 'GetHeight'](element) * ratio)
 	end
 
-	element.casting = nil
-	element.notInterruptible = nil
-	element.holdTime = element.timeToHold or 0
-
-	--[[ Callback: Castbar:PostCastFailed(unit)
-	Called after the element has been updated upon a failed spell cast.
+	--[[ Callback: Castbar:PostCastStart(unit)
+	Called after the element has been updated upon a spell cast start.
 
 	* self - the Castbar widget
-	* unit - unit for which the update has been triggered (string)
+	* unit - the unit for which the update has been triggered (string)
 	--]]
-	if(element.PostCastFailed) then
-		return element:PostCastFailed(unit)
+	if(element.PostCastStart) then
+		element:PostCastStart(unit)
 	end
+
+	element:Show()
 end
 
-local function UNIT_SPELLCAST_INTERRUPTED(self, event, unit, castID)
-	if(self.unit ~= unit and self.realUnit ~= unit) then return end
+local function CastUpdate(self, event, unit, castID, spellID)
+	if(self.unit ~= unit) then return end
 
 	local element = self.Castbar
-	if(element.castID ~= castID) then
+	if(not element:IsShown() or element.castID ~= castID or element.spellID ~= spellID) then
```

Comment:
Btw is the `element:IsShown()` check really needed?

Reply:
The `:IsShown()` checks are needed since those events fire for crafting spells as well, in those cases both `castID` and `spellID` will be nil. Looks like I actually missed one in `CastStop`

The absence of `castID`s in channels' events' payloads is the reason why I compare `spellID`s instead, we don't want to accidentally updated a bar for some other channelled ability. It's unlikely to happen, but better safe than sorry.

And Blizz code hasn't been updated after they reworked events' payloads, so it's a bad reference.

This comment applies to all of your comments as well.